### PR TITLE
Add proxy.skip_tls_verify config option

### PR DESCRIPTION
**Pull request by n-oden**
## Description

If `server.proxy.skip_tls_verify` is set to true, create a custom transport for the reverse proxy that disables TLS verification; this would be used e.g. for clusters that have self-signed certificates for testing.

## Pull request type

Please check the type of change your PR introduces:
- [ ] Bugfix
- [X] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

### Checklist

- [X] Linter passes correctly
- [X] Add tests which fail without the change (if possible)
- [X] All tests passing
- [X] Extended the README / documentation, if necessary

## Does this introduce a breaking change?

- [ ] Yes
- [X] No

### CODE REVIEW

File: main.go
Comment (mga-chka) [inline main.go]: what is the link with the skip tls feature?

File: main.go
Comment (mga-chka) [inline main.go:286]: since the Proxy part is optional in the configuration, this line looks suspicious.

File: docs/content/en/configuration/default.md
Comment (mga-chka) [inline docs/content/en/configuration/default.md:169]: thanks for adding a bit of comment in the doc so that other people are aware of this feature.
Can you also add a brief description in this file chproxy/docs/content/en/configuration/proxy.md ?
Because this is what people see in the proxy part on chproxy.org => https://www.chproxy.org/configuration/proxy

File: main.go
Comment (n-oden) [reply main.go]: Oops, sorry, I meant to put this in the PR description but was rushing out the door and forgot.  It's not linked to the skip_tls feature but was definitely something we found useful internally when running under kubernetes.  I can flesh it out here or open a different PR if you'd prefer.

Comment (mga-chka) [reply main.go]: yes, it's better to put it in another PR. FYI we're also using chproxy under kube and we don't need the /health endpoint

File: main.go
Comment (n-oden) [reply main.go:286]: Okay, I took another pass at this, but I'm open to feedback here: It's not entirely clear to me how dynamic you _want_ this behavior to be?  The current tests in main_test don't reset server state between them so I tried to make the behavior work unsurprisingly with that.

File: main.go
Comment (n-oden) [reply main.go]: 👍

File: docs/content/en/configuration/default.md
Comment (n-oden) [reply docs/content/en/configuration/default.md:169]: Done!

File: main.go
Comment (mga-chka) [reply main.go:286]: my bad I wasn't explicit enough. Since is Proxy is optional in the conf, my guess is that cfg.Server.Proxy can be nil so using cfg.Server.Proxy.SkipTLSVerify could lead to some crashes. So it better to check if cfg.Server.Proxy is not nil before using cfg.Server.Proxy.SkipTLSVerify. I'm not 100% sure but it's worth digging

Comment (n-oden) [reply main.go:286]: I _think_ that the current check for this using reflect.ValueOf().IsZero is correct: as long as you initialize the config struct the inner structs should be there (non-nil) with their zero values -- if that weren't the case I'd expect `TestReloadConfig` to panic on L291 or L298 since `testdata/http.yml` doesn't have a proxy section.

Comment (n-oden) [reply main.go:286]: I've re-arranged the logic in applyConfig to make it clearer what's going on.

Comment (mga-chka) [reply main.go:286]: my bad,
I just checked the code of config.go and even if the Proxy is optional in the yaml, it's always present in the struct since it has this form
```yaml
type Server struct {
[...]
	// Optional Proxy configuration
	Proxy Proxy `yaml:"proxy,omitempty"`
[...]
}
```

I don't know why but I thought the struct had this form (and since it was the weekend I just  looked at your PR)
```yaml
type Server struct {
[...]
	// Optional Proxy configuration
	Proxy *Proxy `yaml:"proxy,omitempty"`
[...]
}
```
So, despite what I said in my first comment,  you can rollback to the first version of the applyConfig since it's easier to read and safe (sorry about that).

Comment (n-oden) [reply main.go:286]: No worries -- done!

**Review by mga-chka — APPROVED** [review]
Looks good to me.
Please fix the lint issue so that I can merge it.

After your change, I'll wait a bit before merging the PR to let other maintainers looking at the PR if they want.
This feature will be part of the future 1.20.0

**Comment by n-oden** [comment]
Sigh, I just realized that this PR doesn't actually touch the heartbeat code, which it's definitely going to have to; look for an update tomorrow.

File: main.go
Comment (n-oden) [reply main.go]: Belated followup question: when you run chproxy under k8s are you using a livenessProbe or a readinessProbe and if so what URL do you use for that?

**Comment by n-oden** [comment]
@mga-chka I think this is good to go?

**Comment by mga-chka** [comment]
> @mga-chka I think this is good to go?

I'm off, I'll look at it on monday

**Comment by mga-chka** [comment]
@n-oden why do you need to by pass the TLS verification on both the heartbeat and the delete queries? Because from what I understood, your aim was to bypass TLS between the reverse proxy and chproxy. But with your last commit the reverse proxy is not involved

**Comment by n-oden** [comment]
My goal here is to allow an option to skip TLS verification between chproxy and clickhouse itself.  The specific use cases would be a clickhouse server that is using a self-signed cert or a cert signed by a private CA on port 9000, which one might do for testing or internal connections to an Altinity Cloud Managed clickhouse server (they use a private CA for their "edge proxy" service).

**Comment by mga-chka** [comment]
Sorry for the time to respond but I spent a bit of time to think about this PR since I misunderstood a bit the aim of the PR.
 
In config.go, the SkipTLSVerify field should not be inside the type Proxy because this struct is linked to anything related to having a reverse proxy before chproxy (like a load balancer).
Theoretically it should be in the type Cluster  (just after the Scheme field that defines if the traffic is https or not) because you want to bypass TLS just for clickhouse. But doing that would require some refactoring so put it in type Config struct (for example after the HackMePlease field). 

In default.md you should be more explicit that its for the TLS validation of the clickhouse servers in https (for example it could be for redis and soon chproxy will provide a tcp connexion for clickhouse).
You can also remove the description in proxy.md since it's not linked to the proxy part.

File: main.go
Comment (mga-chka) [reply main.go]: I'll let @sigua-cs answer since I didn't worked on the kube part but I think we're using the /ping

Comment (sigua-cs) [reply main.go]: @n-oden in our statefulset manifest, we use `/metrics` endpoint for both probes
```
livenessProbe:
  httpGet:
    path: /metrics
    port: http
readinessProbe:
  httpGet:
    path: /metrics
    port: http
```

**Comment by n-oden** [comment]
@mga-chka all reasonable feedback; I'll try to get to this early next week

**Comment by mga-chka** [comment]
@n-oden any news regarding the improvements (because we're waiting for this PR to release 1.20.0)?

**Comment by mga-chka** [comment]
@n-oden I'm just pinging you in case you forgot the previous one because another PR (275) is on hold since it's linked to this one

**Comment by mga-chka** [comment]
I'm closing this PR since I didn't get any update. Feel free to re-open it if you want to finish it.

**Pull request closed without merging**
